Android/app/src/main/java/app/intra/ServerChooserFragment.java

```diff
@@ -0,0 +1,155 @@
+        final Bundle bundle = new Bundle(1);
+        bundle.putString(ARG_KEY, key);
+        fragment.setArguments(bundle);
+        return fragment;
+    }
+
+    @Override
+    public void onCheckedChanged(RadioGroup group, int checkedId) {
+        if (checkedId == R.id.pref_server_google) {
+            newUrl = getResources().getString(R.string.url0);
+            setValid(true);
+            text.setEnabled(false);
+        } else if (checkedId == R.id.pref_server_cloudflare) {
+            newUrl = getResources().getString(R.string.url1);
+            setValid(true);
+            text.setEnabled(false);
+        } else {
+            newUrl = text.getText().toString();
+            setValid(checkUrl(Untemplate.strip(newUrl)));
+            text.setEnabled(true);
+        }
+    }
+
+    @Override
+    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+    }
+
+    @Override
+    public void onTextChanged(CharSequence s, int start, int before, int count) {
+    }
+
+    @Override
+    public void afterTextChanged(Editable s) {
+        newUrl = s.toString();
+        setValid(checkUrl(Untemplate.strip(newUrl)));
+    }
+
+    // Check that the URL is a plausible DOH server: https with a domain, a path (at least "/"),
+    // and no query parameters or fragment.
+    private boolean checkUrl(String url) {
+        try {
+            URL parsed = new URL(url);
+            return parsed.getProtocol().equals("https") && !parsed.getHost().isEmpty() &&
+                !parsed.getPath().isEmpty() && parsed.getQuery() == null && parsed.getRef() == null;
+        } catch (MalformedURLException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
+        // Usability optimization:
+        // If the user is typing in the free text field and presses "enter" or "go" on the keyboard
+        // while the URL is valid, treat that the same as closing the keyboard and pressing "OK".
+        if (checkUrl(Untemplate.strip(v.getText().toString()))) {
+            Dialog dialog = getDialog();
+            if (dialog instanceof AlertDialog) {
+                Button ok = ((AlertDialog)dialog).getButton(AlertDialog.BUTTON_POSITIVE);
+                ok.setEnabled(true);
+                ok.performClick();
+            }
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected void onBindDialogView(View view) {
+        super.onBindDialogView(view);
+        ServerChooser preference = (ServerChooser) getPreference();
+        String url = preference.getUrl();
+        buttons = view.findViewById(R.id.pref_server_radio_group);
+        text = view.findViewById(R.id.custom_server_url);
+        if (url == null || url.equals(getResources().getString(R.string.url0))) {
+            buttons.check(R.id.pref_server_google);
+            text.setEnabled(false);
```

Comment:
And this

Reply:
In this design, I can't call `updateUI` here because `setValid` requires `Dialog` to exist, but it doesn't exist yet.

Reply:
Or you could do a null check?  Anyway, your choice :)

Reply:
I switched this to a more MVC-ish design.  Using `onCreateView` as we discussed doesn't work, because the action buttons don't exist yet, but I found a later life cycle event that works.